### PR TITLE
feat(Customer Center): load full purchase history and gate the entry point (2/4)

### PR DESCRIPTION
**Pull request by facumenzella**
> **Stacked PR 2 of 4** (merge in order). Supersedes #3398.
>
> 1. #3997 config + model fields (175 lines)
> 2. #3998 loading + state + gate (99 lines)
> 3. #3999 history list screen (284 lines)
> 4. #4000 detail screen (295 lines)

## Summary

Stack 2 of 4 for Customer Center Purchase History. Adds the loading path and visibility gate, still no UI.

- `loadAllPurchases` groups purchases into active subs, inactive subs and non-subs, each sorted by purchase date, matching iOS `PurchaseHistoryViewModel`. Products resolve concurrently.
- `loadPurchases` now receives the `CustomerInfo` it used to fetch itself, so one `FETCH_CURRENT` feeds both paths instead of two.
- `allPurchases` is only populated when `display_purchase_history_link` is on.
- `shouldShowPurchaseHistory` ports iOS `shouldShowSeeAllPurchasesButton` exactly, threshold of 2 included.

**Production churn: 99 / 300.**

<details>
<summary>AI session context</summary>

## Metadata

- Branch: see PR head
- Author / human owner: facumenzella
- Agent(s): Claude Code (Opus 5, 1M context)
- Session source: current conversation
- Generated: 2026-08-18
- Context document version: 1

## Goal

Land the Customer Center Purchase History feature (iOS parity) as four small stacked PRs instead of one 863-line PR. Supersedes #3398.

## Initial Prompt

"Check https://github.com/RevenueCat/purchases-android/pull/3398. Make it up-to-date with main, and let's see if we're missing anything" (the actionable task: merge main into #3398, verify it, and report parity/CI gaps).

## Important Follow-up Prompts

- "how do you feel about splitting this pr into smaller pieces? Stacked" / "Maybe 2 prs or 3 if it's simpler", which redirected the work from "update #3398" to "replace it with a stack".

## Agent Contribution

- Merged `origin/main` into #3398 (377 commits behind, merge base 2026-05-05). Clean merge, zero conflicts.
- Verified the merge rather than trusting green CI: inspected all 5 deletions in the diff vs main and confirmed each was the PR's own refactor, and confirmed main's only Customer Center change in that window (`presentedOfferingContext = offering.presentedOfferingContext`) survived.
- Diagnosed the new `revenuecat/danger` failure as main's added 300-line production-churn rule, not a code regression. Measured #3398 at 863 lines.
- Compared against `purchases-ios` `origin/main` and found the localization keys, section titles, grouping/sorting, and the `shouldShowSeeAllPurchases` gate all match, plus one divergence (debug rows, see below).
- Carved the verified merged state into four stacked branches, compiling and testing each layer independently.

## Human Decisions

- Decision: 4 stacked PRs, no Danger bypass label on any of them.
- Decision: close #3398 and link the stack rather than rebase it into the stack.
- Decision: leave the iOS debug-rows divergence as-is and document it rather than change behavior.

## Key Implementation Decisions

- Decision: merge `main` into #3398 rather than rebase.
  - Rationale: the branch already carried four `Merge branch 'main'` commits; rebasing 15 commits across 377 would replay conflicts per commit.
  - Rejected: `git rebase origin/main` (the repo's usual cold-start step).
- Decision: split by navigation depth, not by screen file.
  - Rationale: neither the `CustomerCenterAction` nor `CustomerCenterDestination` `when` has an `else`, and Kotlin 2.0 makes a non-exhaustive `when` over a sealed subject a compile error. A destination cannot land before the view its nav branch renders.
  - Rejected: both destinations in one PR (would force both screens into it, about 530 lines).
- Decision: `PurchaseHistoryView` has no `onAction` parameter until PR 4.
  - Rationale: avoids shipping an unused parameter in PR 3.

## Validation

- Commands run (per layer):
  - `./gradlew :purchases:compileDefaultsDebugKotlin :ui:revenuecatui:compileDefaultsDebugKotlin`: BUILD SUCCESSFUL on every branch, each compiled standalone on its own base.
  - `./gradlew :ui:revenuecatui:testDefaultsDebugUnitTest`: BUILD SUCCESSFUL. Full suite run on all four branches, not just the tip.
  - `./gradlew :purchases:testDefaultsDebugUnitTest detektAll`: BUILD SUCCESSFUL.
  - `./scripts/api-check.sh`: exit 0. No `api-*.txt` change needed, `CustomerCenterConfigData` is `@InternalRevenueCatAPI` so Metalava excludes it.
- Manual verification:
  - Stack tip reconstructs the verified merged production code byte-for-byte in 12 of 13 files. The 13th, `PurchaseHistoryDetailView.kt`, differs only by the intentional preview simplification described in PR 4.
  - Each of the 8 purchase-history view model tests from #3398 appears exactly once across the stack.
  - No app run, no device or emulator verification.
- CI:
  - #3998: 34 checks pass, `codecov/project` still pending at time of writing.

## Validation Gaps

- No Paparazzi snapshot coverage in-repo for the new screens. Their rendered output is covered by Emerge snapshot testing in CI instead (see #3999 and #4000).
- `PaywallComponentsTemplatePreviewRecorder` fails in a fresh worktree until `upstream/paywall-preview-resources` is initialized. Confirmed environmental: it passes after `git submodule update --init`.
- No verification against a real backend payload for `display_purchase_history_link`.

## Risks / Reviewer Notes

- Risk: iOS gates `store`, `productID`, `sandbox`, `transactionID` and `originalPurchaseDate` behind a `#if DEBUG` section. This port renders four of those five inline and unconditionally, and has no `store` row or `Debug` header.
  - Evidence: iOS `PurchaseDetailItem.isDebugOnly` and `purchaseDetailDebugItems`. Cursor flagged the rows as missing on #3398 and the thread was resolved by adding them, which moved the port away from the source of truth.
  - Mitigation: none applied. `#if DEBUG` has no 1:1 Android equivalent, since a prebuilt AAR's `BuildConfig.DEBUG` is the library's build type rather than the consuming app's. Matching the intent would need something like `ApplicationInfo.FLAG_DEBUGGABLE`, which is a product decision.
- Risk: when `display_purchase_history_link` is on, `loadAllPurchases` runs on every Customer Center load and resolves products for active Play Store subscriptions.
  - Evidence: gated on the flag, so projects without the feature pay nothing. Product lookups run concurrently, not serially.
  - Mitigation: reviewer should sanity-check behavior for customers with large purchase histories.

## Non-goals / Out of Scope

- Changing the debug-row behavior.
- Adding snapshot coverage for the new screens.
- Any change to #3398 itself beyond the main merge.

## Omitted Context

- Raw transcript, unrelated exploration, sensitive details, repetitive attempts, and chain-of-thought-style content were omitted.

</details>

---

> [!NOTE]
> **Medium Risk**
> When purchase history is enabled, every Customer Center load may fetch full history and run concurrent product lookups for active Play Store subs; grouping logic must stay aligned with iOS to avoid wrong sections or a hidden “see all” link.
> 
> **Overview**
> Adds the **data and loading layer** for Customer Center purchase history (stack 2/4); there is still **no UI** for the history screen or link in this PR.
> 
> **State** now carries grouped `purchaseHistory` (active subs, inactive subs, non-subs) and `shouldShowPurchaseHistory`. Grouping uses raw `CustomerInfo` subscription/non-subscription collections—not `PurchaseInformation` display flags—so promotional subs and entitlement-backed “active” rows land in the right buckets.
> 
> **Loading**: one `FETCH_CURRENT` `CustomerInfo` feeds both the existing relevant-purchases list and optional full history when `displayPurchaseHistoryLink` is true. `loadPurchaseHistory` sorts by purchase date and resolves Play Store products for active subs **concurrently** via `async`/`awaitAll`.
> 
> **Visibility gate**: `CustomerInfo.shouldShowSeeAllPurchases()` mirrors iOS (mixed active/inactive subs, multiple inactive-only subs, or more than two non-sub transactions), combined with the config flag for `shouldShowPurchaseHistory`.
> 
> Unit tests cover history population, field mapping, entry IDs, grouping edge cases, and link visibility.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c7231fbd27ccde3ed8a2fe5faf3b942e82e6b392. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerCenterState.kt [hunk 1]
```diff
@@ -28,6 +28,7 @@ internal sealed class CustomerCenterState(
     data class Success(
         @get:JvmSynthetic val customerCenterConfigData: CustomerCenterConfigData,
         @get:JvmSynthetic val purchases: List<PurchaseInformation> = emptyList(),
+        @get:JvmSynthetic val allPurchases: List<PurchaseInformation> = emptyList(),
```
Comment (vegaro) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerCenterState.kt]: this `purchases` vs `allPurchases` is not super clear

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerInfoSeeAllPurchases.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Returns true if:
+ * - There are both active and inactive subscriptions.
+ * - There are only inactive subscriptions and more than one of them.
+ * - The number of non-subscription purchases exceeds [maxNonSubscriptions].
+ *
+ * This counts the raw customer info collections rather than the mapped [PurchaseInformation] list.
+ * `isSubscription` and `isExpired` are display concerns (`isSubscription` excludes promotional
+ * purchases, `isExpired` follows entitlement activity), so grouping by them undercounts here.
+ */
+internal fun CustomerInfo.shouldShowSeeAllPurchases(
```
Comment (vegaro) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerInfoSeeAllPurchases.kt:23]: this is very clear and easy to follow thanks

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/viewmodel/CustomerCenterViewModel.kt [hunk 3]
```diff
@@ -767,6 +770,74 @@ internal class CustomerCenterViewModelImpl(
         }
     }
 
+    private suspend fun loadAllPurchases(
```
Comment (vegaro) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/viewmodel/CustomerCenterViewModel.kt]: again, this `loadPurchases` vs `loadAllPurchases` is not very clear

**Review by vegaro — APPROVED** [review]

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerCenterState.kt (hunk 1 above)
Comment (facumenzella) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/data/CustomerCenterState.kt]: Renamed to `purchaseHistory` in https://github.com/RevenueCat/purchases-android/commit/a732a757cef4114f4cd6e240403111d6eaf23501. Left `purchases` alone, it predates this PR and it is not a safe mechanical rename in this package.

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/viewmodel/CustomerCenterViewModel.kt (hunk 3 above)
Comment (facumenzella) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/customercenter/viewmodel/CustomerCenterViewModel.kt]: Same commit, it is `loadPurchaseHistory` now: https://github.com/RevenueCat/purchases-android/commit/a732a757cef4114f4cd6e240403111d6eaf23501
